explorer/interpreter/resolve_names.cpp

```diff
@@ -273,15 +273,30 @@ auto NameResolver::ResolveNames(Expression& expression,
       break;
     }
     case ExpressionKind::SimpleMemberAccessExpression: {
+      // If the left-hand side of the `.` is a namespace or alias to namespace,
+      // resolve the name.
       auto& access = cast<SimpleMemberAccessExpression>(expression);
       CARBON_ASSIGN_OR_RETURN(std::optional<ValueNodeView> scope,
                               ResolveNames(access.object(), enclosing_scope));
+      if (!scope) {
+        break;
+      }
 
-      // If the left-hand side of the `.` is a namespace, resolve the name.
-      // TODO: Look through aliases.
-      if (scope && isa<NamespaceDeclaration>(scope->base())) {
-        auto ns_it =
-            namespace_scopes_.find(&cast<NamespaceDeclaration>(scope->base()));
+      Nonnull<const AstNode*> base = &scope->base();
+      // recursively resolve aliases
+      auto namespace_alias =
+          [](const AstNode& base) -> std::optional<Nonnull<const AstNode*>> {
+        if (const auto* alias = dyn_cast<AliasDeclaration>(&base)) {
+          return alias->resolved_declaration();
+        }
+        return std::nullopt;
+      };
+      while (auto new_base = namespace_alias(*base)) {
+        base = *new_base;
+      }
```

Comment:
this feels nicer to me.

but an alternative:

```cpp
while (const auto* alias = dyn_cast<AliasDeclaration>(base)) {
  if (auto resolved = alias->resolved_declaration()) {
    base = *resolved;
  } else {
    break;
  }
}
```

Reply:
I'd prefer the alternative approach; the use of a lambda adds more flow control paths to understand, whereas the `while`/`break` approach seems simpler. But also, looking at the alternative I see a simplification that I've suggested (separately, so that I can suggest the full replacement).

Reply:
I have switched to the alternative.